frontend/src/metabase/query_builder/components/notebook/lib/steps.ts

```diff
@@ -94,8 +96,15 @@ const STEPS: NotebookStepDef[] = [
       query.hasData() &&
       (!query.hasAggregations() || query.hasBreakouts()) &&
       (!query.sourceQuery() || query.hasAnyClauses()),
-    active: query => query.hasLimit(),
-    revert: query => query.clearLimit(),
+    active: (legacyQuery, itemIndex, query, stageIndex) =>
+      Lib.hasLimit(query, stageIndex),
+    revert: (legacyQuery, itemIndex, query, stageIndex) => {
+      const reverted = Lib.limit(query, stageIndex, null);
+      const revertedLegacyQuery = Lib.toLegacyQuery(reverted);
+      const revertedStructuredQuery =
+        legacyQuery.setDatasetQuery(revertedLegacyQuery);
+      return revertedStructuredQuery.queries()[stageIndex];
+    },
```

Comment:
I don't love making notebook step functions receive four args now, but that felt like the most sane way to keep both v1 and v2 patterns in place